drakrun/drakrun/main.py

```diff
@@ -356,187 +405,182 @@ def build_drakvuf_cmdline(self, timeout, cwd, full_cmd, dump_dir, ipt_dir, workd
-            self.log.error("Tried to run the analysis for more than hard limit of %d seconds", hard_time_limit)
+            self.log.error(
+                "Tried to run the analysis for more than hard limit of %d seconds",
+                hard_time_limit
+            )
             return
 
-        self.log.info(f"analysis UID: {self.analysis_uid}")
-        self.backend.redis.set(f"drakvnc:{self.analysis_uid}", self.instance_id, ex=3600)  # 1h
+        self.update_vnc_info()
 
-        workdir = f"/tmp/drakrun/{self.vm_name}"
-
-        extension = self.current_task.headers.get("extension", "exe").lower()
+        # Get sample extension. If none set, fall back to exe/dll
+        extension = task.headers.get("extension", "exe").lower()
         if '(DLL)' in magic_output:
             extension = 'dll'
         self.log.info("Running file as %s", extension)
 
-        file_name = self.current_task.payload.get("file_name", "malwar") + f".{extension}"
+        # Prepare sample file name
+        file_name = task.payload.get("file_name", "malwar") + f".{extension}"
         # Alphanumeric, dot, underscore, dash
         if not re.match(r"^[a-zA-Z0-9\._\-]+$", file_name):
             self.log.error("Filename contains invalid characters")
             return
         self.log.info("Using file name %s", file_name)
 
-        try:
-            shutil.rmtree(workdir)
-        except OSError:
-            self.log.exception("Failed to clean work directory")
+        # workdir - configs, sample, etc.
+        # outdir - analysis artifacts
+        workdir, outdir = self._prepare_workdir()
 
-        os.makedirs(workdir, exist_ok=True)
+        sample_path = os.path.join(workdir, file_name)
+        sample.download_to_file(sample_path)
+
+        # Try to come up with a start command for this file
+        # or use the one provided by the sender
+        cmd = self._get_start_command(
+            extension,
+            sample,
+            sample_path
+        )
 
-        # Save sample to disk here as some branches of _get_start_command require file path.
-        with open(os.path.join(workdir, file_name), 'wb') as f:
-            f.write(sample.content)
+        start_command = task.payload.get("start_command", cmd)
+        if not start_command:
+            self.log.error(
+                "Unable to run malware sample. Could not generate any suitable"
+                " command to run it."
+            )
+            return
+        self.log.info("Start command: %s", start_command)
 
         # If task contains 'custom_hooks' override local defaults
         with open(os.path.join(workdir, "hooks.txt"), "wb") as hooks:
-            if self.current_task.has_payload("custom_hooks"):
-                custom_hooks = self.current_task.get_resource("custom_hooks")
+            if task.has_payload("custom_hooks"):
+                custom_hooks = task.get_resource("custom_hooks")
+                assert custom_hooks.content is not None
```

Comment:
There's a typing in https://github.com/CERT-Polska/karton/blob/master/karton/core/resource.py#L390
which allows the `content` to be None. It doesn't seem to be correct (it doesn't seem reasonable to me for a resource to have no content, if it doesn't why is there a resource in the first place).
This triggers warning from static type checker since from their point of fiew `custom_hooks.content` can be None

Reply:
Should be fixed in upstream, but I guess this is ok here.